etl/src/v2/concurrency/future.rs

```diff
@@ -0,0 +1,294 @@
+use futures::future::{BoxFuture, CatchUnwind};
+use futures::FutureExt;
+use pin_project_lite::pin_project;
+use std::future::Future;
+use std::panic::AssertUnwindSafe;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use tokio::sync::RwLock;
+
+/// A callback interface for receiving notifications about future completion.
+///
+/// The generic parameter `I` represents the identifier type used to track futures.
+pub trait ReactiveFutureCallback<I> {
+    /// Called when a future completes successfully.
+    fn on_complete(&mut self, id: I);
+
+    /// Called when a future encounters an error.
+    fn on_error(&mut self, id: I, error: String);
+}
+
+/// Represents the internal state of a [`ReactiveFuture`].
+enum ReactiveFutureState {
+    /// The inner future is currently being polled.
+    PollInnerFuture,
+    /// The inner future has completed, and we have to notify the callback.
+    InvokeCallback(Option<String>),
+    /// The future has finished all processing.
+    Finished,
+}
+
+pin_project! {
+    /// A future that notifies a callback when it completes or encounters an error.
+    ///
+    /// The generic parameters are:
+    /// - `Fut`: The underlying future type
+    /// - `I`: The identifier type for tracking futures
+    /// - `C`: The callback type that implements [`ReactiveFutureCallback`]
+    pub struct ReactiveFuture<Fut, I, C> {
+        #[pin]
+        future: CatchUnwind<AssertUnwindSafe<Fut>>,
```

Comment:
I'm a little concerned that using `AssertUnwindSafe` to wrap a full future might lead to logic bugs if the future panics and some invariants are broken. Not sure if we can somehow avoid this in generic code though. The other option is to treat any kind of panic as a bugs to be fixed, and rely only on tasks returning results via `JoinHandle`s. We should anyway be prepared for the process crashes to gracefully resume replication.